### ISSUE TITLE
make install command is not working from git bash

### ISSUE THREAD

**Issue by garimakaushik22** [DAI-Lab/SteganoGAN#23]
I cloned the repository  from GitHub and as given in readme tried to execute make install command.
I am getting below error. 

$ make install
bash: make: command not found

**Comment by csala** [comment]
Hello @garimakaushik22 Would you mind providing some more details about you environment?

Specifically, can you please fill the following form?

* SteganoGAN version or git commit:
* Python version:
* Operating System:

### Description

Describe what you were trying to get done.
Tell us what happened, what went wrong, and what you expected to happen.

### What I Did

```
Paste the command(s) you ran and the output.
If there was a crash, please include the traceback here.
```

Anyway, the error that you see seems to indicate that you don't have `make` installed, and, from what you say at your other issue, it seems like you are running windows, so I suggest you to search how to "Install make on windows", do it and try again.

**Comment by garimakaushik22** [comment]
Hello,
Yes I am using Windows 10. May I know what environment is this coded in? Is their any specific requirement needs to be taken care when running in windows.

•Python version (output of pip --version): Python 3.6.5
•SteganoGAN version or git commit:  -current_version = 0.1.1
+current_version = 0.1.2-dev

•Operating System: Windows 10

**Comment by csala** [comment]
Hi @garimakaushik22 
The only requirement is that your system is able to build python packages using `make`, which in most cases is just a matter of installing it.
Regarding the platform, this was coded in Mac and Linux machines, but it should not be relevant in this case, since in most cases `make` is not there by default in mac and linux and also needs to be installed manually beforehand.
Were you able to install `make` and try?

**Comment by garimakaushik22** [comment]
Hi,
Sorry for the late reply. I was able to install make command but I am still getting error running make install.

Error:
"fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument 'HEAD..stable': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
rm -fr build/
rm -fr dist/
rm -fr .eggs/
find . -name '*.egg-info' -exec rm -fr {} +
find . -name '*.egg' -exec rm -f {} +
find . -name '*.pyc' -exec rm -f {} +
find . -name '*.pyo' -exec rm -f {} +
find . -name '*~' -exec rm -f {} +
find . -name '__pycache__' -exec rm -fr {} +
pip install .
make: *** [install] Error 1
"

**Comment by csala** [comment]
Hello @garimakaushik22 I'm afraid we need more details about the error and its context.

Can you paste all the details? Right now we can only see that there was an error, but it's hard to see where it came from.

Also, would you mind formatting it using markdown? Here you have a guide: https://help.github.com/articles/creating-and-highlighting-code-blocks/

**Comment by garimakaushik22** [comment]
I am running make install command from gitbash after cloning the repository.
Please find the complete stack trace. I have installed make
```
garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~
$ cd SteganoGAN

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/SteganoGAN (master)
$ make install
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument 'HEAD..stable': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
rm -fr build/
rm -fr dist/
rm -fr .eggs/
find . -name '*.egg-info' -exec rm -fr {} +
find . -name '*.egg' -exec rm -f {} +
find . -name '*.pyc' -exec rm -f {} +
find . -name '*.pyo' -exec rm -f {} +
find . -name '*~' -exec rm -f {} +
find . -name '__pycache__' -exec rm -fr {} +
pip install .
make: *** [install] Error 1

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/SteganoGAN (master)

```

**Comment by pvk-developer** [comment]
Hi!
The problem there is that you may be missing the branch `stable`. In order to fix this, just run:
`git branch stable`. This should allow you to use `make`.

**Comment by csala** [comment]
Hi @garimakaushik22 

We looked at your error trace and tried to reproduce the error, and we cannot see what it could be, as the trace provides no information.

Can you please try to execute the following and report back?

```
pip install .
```

Just to clarify, this is just what `make install` is running under the hoods, so hopefully we will obtain more information out of it.

**Comment by garimakaushik22** [comment]
Hi @csala 

Please find the stack trace of error

```
garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~
$ cd SteganoGAN

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/SteganoGAN (master)
$ pip install

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/SteganoGAN (master)
$ steganogan encode C:\\Users\\g.a.kaushik\\Desktop\\test\\demo.png  "secret"
bash: steganogan: command not found

```

**Comment by csala** [comment]
Hello @garimakaushik22 

Please copy the entire command. Notice that there is a `.` at the end, after the `install`.

Also, it doesn't look like you are using any kind of environment isolation tool.
I recommend you that you the installation inside either a virtualenv or a conda environment.

**Comment by garimakaushik22** [comment]
I am using gitbash for cloning from git and using the command from gitbash itself. 
Error from Git Bash: 
```
garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~
$ cd C:\Users\g.a.kaushik\Desktop
bash: cd: C:Usersg.a.kaushikDesktop: No such file or directory

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~
$ cd C:\\Users\\g.a.kaushik\\Desktop

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/Desktop
$ cd SteganoGAN

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/Desktop/SteganoGAN (master)
$ pip install .

garima.g.kaushik@BDC7-L-2C2QKC2 MINGW64 ~/Desktop/SteganoGAN (master)
$ steganogan encode C:\\Users\\g.a.kaushik\\Desktop\\test\\demo.png  "secret"
bash: steganogan: command not found

```

I have tried the same thing from anaconda prompt and getting below error:

```

(steg) C:\Users\g.a.kaushik\SteganoGAN>pip install .
Processing c:\users\g.a.kaushik\steganogan
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\GA07C1~1.KAU\AppData\Local\Temp\pip-req-build-_hd0cfgh\setup.py", line 9, in <module>
        readme = readme_file.read()
      File "C:\Anaconda\envs\steg\lib\encodings\cp1252.py", line 23, in decode
        return codecs.charmap_decode(input,self.errors,decoding_table)[0]
    UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 131: character maps to <undefined>

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in C:\Users\GA07C1~1.KAU\AppData\Local\Temp\pip-req-build-_hd0cfgh\

(steg) C:\Users\g.a.kaushik\SteganoGAN>

```

**Comment by csala** [comment]
@garimakaushik22 Please either use a conda environment, or make sure to create a virtualenv with the right python versions in it in the other one.

Then, once you are in the conda prompt, that error showed up because there were two characters in the README that windows encoding does not support. I already replaced them, so please pull from `master` and try again.

**Comment by garimakaushik22** [comment]
Hi @csala

It worked in anaconda prompt. But make install still failed whereas pip install worked fine. 
I am getting the encoded/ decoded output both. But I am getting some warning after the output.

```
(steg) C:\Users\g.a.kaushik\Desktop\test1\SteganoGAN>steganogan encode -o C:\\Users\\g.a.kaushik\\Desktop\\test\\out.png  C:\\Users\\g.a.kaushik\\Desktop\\test\\demo.png "Message to hide"
c:\anaconda\envs\steg\lib\site-packages\torch\serialization.py:434: SourceChangeWarning: source code of class 'steganogan.decoders.DenseDecoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
c:\anaconda\envs\steg\lib\site-packages\torch\serialization.py:434: SourceChangeWarning: source code of class 'steganogan.critics.BasicCritic' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)

```

(steg) C:\Users\g.a.kaushik\Desktop\test1\SteganoGAN>steganogan decode C:\\Users\\g.a.kaushik\\Desktop\\test\\out.png
c:\anaconda\envs\steg\lib\site-packages\torch\serialization.py:434: SourceChangeWarning: source code of class 'steganogan.decoders.DenseDecoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
c:\anaconda\envs\steg\lib\site-packages\torch\serialization.py:434: SourceChangeWarning: source code of class 'steganogan.critics.BasicCritic' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
Message to hide

```

**Comment by csala** [comment]
@garimakaushik22 Let's keep this issue only for the `make install` issue.

If you were able to execute `make install` successfully and errors show up when running the `steganogan` command, please close this issue and open a new one reporting the error there.

**Comment by garimakaushik22** [comment]
@csala

make install didn't work. I was able to use pip install.
As mentioned in my last comment  why am I getting warning message? 

```
(steg) C:\Users\g.a.kaushik\Desktop\test1\SteganoGAN>steganogan decode C:\Users\g.a.kaushik\Desktop\test\out.png
 c:\anaconda\envs\steg\lib\site-packages\torch\serialization.py:434: SourceChangeWarning: source code of class 'steganogan.decoders.DenseDecoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set torch.nn.Module.dump_patches = True and use the patch tool to revert the changes.
 warnings.warn(msg, SourceChangeWarning)
 c:\anaconda\envs\steg\lib\site-packages\torch\serialization.py:434: SourceChangeWarning: source code of class 'steganogan.critics.BasicCritic' has changed. you can retrieve the original source code by accessing the object's source attribute or set torch.nn.Module.dump_patches = True and use the patch tool to revert the changes.
 warnings.warn(msg, SourceChangeWarning)
 Message to hide

```

**Comment by csala** [comment]
Hello @garimakaushik22 

You are able to execute the `steganogan` command only after a successful installation, so we can assume that the `make install` command worked properly.

Regarding the warnings, this is the expected behavior when using the `master` branch, as the pretrained models correspond to an older version of the code. In order too avoid those, please checkout the `stable` branch.

But, in any case, the warnings do not mean that the command didn't work: you should be able to still encode and decode messages despite the warnings.
If you aren't, please open a new issue reporting the corresponding error and we will follow up there.
